eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/ImpTest.hs

```diff
@@ -516,25 +517,25 @@ submitPhase2Invalid tx = do
   impAnn "Submit tx with IsValid False" $ do
     withNoFixup $ submitTx $ fixedUpTx & isValidTxL .~ IsValid False
 
-expectTxSuccess ::
+impAlonzoExpectTxSuccess ::
   ( HasCallStack
   , AlonzoEraImp era
   ) =>
   Tx era -> ImpTestM era ()
-expectTxSuccess tx
-  | tx ^. isValidTxL == IsValid True = do
-      utxo <- getsNES utxoL
-      let inputs = Set.toList $ tx ^. bodyTxL . inputsTxBodyL
-          outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
+impAlonzoExpectTxSuccess tx = do
+  utxo <- getsNES utxoL
+  let inputs = tx ^. bodyTxL . inputsTxBodyL
+      collaterals = tx ^. bodyTxL . collateralInputsTxBodyL
+      outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
+  if tx ^. isValidTxL == IsValid True
+    then do
       impAnn "Inputs should be gone from UTxO" $
-        expectUTxOContent utxo [(txIn, isNothing) | txIn <- inputs]
+        expectUTxOContent utxo [(txIn, isNothing) | txIn <- Set.toList inputs]
+      impAnn "Collateral inputs should still be in UTxO" $
+        expectUTxOContent utxo [(txIn, isJust) | txIn <- Set.toList $ collaterals \\ inputs]
       impAnn "Outputs should be in UTxO" $
         expectUTxOContent utxo [(txIn, (== Just txOut)) | (txIn, txOut) <- outputs]
-  | otherwise = do
-      utxo <- getsNES utxoL
-      let inputs = tx ^. bodyTxL . inputsTxBodyL
-          collaterals = tx ^. bodyTxL . collateralInputsTxBodyL
-          outputs = Map.toList . unUTxO . txouts $ tx ^. bodyTxL
+    else do
       impAnn "Non-collateral inputs should still be in UTxO" $
         expectUTxOContent utxo [(txIn, isJust) | txIn <- Set.toList $ inputs \\ collaterals]
```

Comment:
I'd like to reuse `impShelleyExpectTxSuccess` but I can't because we need `inputs \\ collaterals` here rather than just `inputs`

Reply:
That is ok, it is just a little duplication, so it is not that bad. 
In fact it is more clear what is going on, because there is a chance that both inputs and collateral use the same UTxO